backend/modules/student/routers/student_instances.py

```
@@ -908,3 +908,125 @@ async def start_simulation_from_instance(
+            raise HTTPException(status_code=404, detail="Simulation instance not found")
+
+        # Verify the simulation is in a completed/graded state (only allow reset if done)
+        if instance.status not in ["completed", "submitted", "graded"]:
+            raise HTTPException(
+                status_code=400,
+                detail=f"Cannot reset a simulation that is not completed. Current status: {instance.status}"
+            )
+
+        # Get simulation_id from cohort assignment
+        if not instance.cohort_assignment:
+            raise HTTPException(status_code=400, detail="Instance is not associated with a cohort assignment")
```

Ruff (0.14.14)

[warning] Abstract `raise` to an inner function

(TRY301)